ledger-service/http-json/src/main/scala/com/digitalasset/http/dbbackend/ContractDao.scala

```diff
@@ -46,12 +46,35 @@ object ContractDao {
   def updateOffset(
       party: domain.Party,
       templateId: domain.TemplateId.RequiredPkg,
-      newOffset: domain.Offset)(implicit log: LogHandler): ConnectionIO[Unit] =
+      newOffset: domain.Offset,
+      lastOffset: Option[domain.Offset])(implicit log: LogHandler): ConnectionIO[Unit] =
     for {
       tpId <- Queries.surrogateTemplateId(
         templateId.packageId,
         templateId.moduleName,
         templateId.entityName)
-      _ <- Queries.updateOffset(party.unwrap, tpId, newOffset.unwrap)
+      rowCount <- Queries.updateOffset(
+        party.unwrap,
+        tpId,
+        newOffset.unwrap,
+        lastOffset.map(_.unwrap))
+      _ <- if (rowCount == 1)
+        fconn.pure(())
+      else
+        fconn.raiseError(StaleOffsetException(party, templateId, newOffset, lastOffset))
```

Comment:
We shouldn't reinvent all this locally; let's let the database do its job, perhaps with a well-placed `SELECT FOR UPDATE`.

Reply:
Ummmm...
```
FOR UPDATE/FOR SHARE Clause. FOR UPDATE causes the rows retrieved by the SELECT statement to be locked as though for update. This prevents them from being modified or deleted by other transactions until the current transaction ends.
```
why do we want to lock them? With optimistic locking (which I thought we agreed we are doing), this problem is usually solved the way I did it, checking for the number or rows updated (usually with a version or timestamp, offset in our case is fine).
With optimistic locking whatever transaction commits first -- wins, the other should retry.

Reply:
in my experience, the less you mess with locking, the better. This is to avoid issues when deadlocked thread started a DB lock but never committed or rolled back the transaction.

Reply:
This isn't so different, as the `update` here also acquires a lock. (And it must, for the compare-and-swap check to be valid.)

Reply:
the duration of this compare-and-swap is as short as the duration of the regular update, It is implemented using atomic `UPDATE .... WHERE last_offset=$lastOffset`. We just check the number of rows returned, which has to be 1, else we re-run the entire flow.
I think this is the simplest and most performant implementation.

Reply:
plus we update offset as the last step in the flow, so any discrepancies would be caught at this point.

Reply:
Yes, and that `update` acquires a lock which is released on commit.